src/lib/statistics/chunk_statistics/histograms/equal_height_histogram.cpp

```diff
@@ -0,0 +1,211 @@
+  auto current_bin_height = size_t{0};
+  for (auto current_end_first_value_idx = size_t{0}; current_end_first_value_idx < value_counts.size();
+       current_end_first_value_idx++) {
+    current_bin_height += value_counts[current_end_first_value_idx].second;
+
+    // Make sure to create last bin.
+    if (current_bin_height >= count_per_bin || current_end_first_value_idx == value_counts.size() - 1) {
+      bin_maximums.emplace_back(value_counts[current_end_first_value_idx].first);
+      bin_distinct_counts.emplace_back(current_end_first_value_idx - current_bin_first_value_idx + 1);
+      current_bin_height = 0ul;
+      current_bin_first_value_idx = current_end_first_value_idx + 1;
+    }
+  }
+
+  return {bin_maximums, bin_distinct_counts, min, total_count};
+}
+
+template <typename T>
+std::shared_ptr<EqualHeightHistogram<T>> EqualHeightHistogram<T>::from_segment(
+    const std::shared_ptr<const BaseSegment>& segment, const BinID max_bin_count,
+    const std::optional<std::string>& supported_characters, const std::optional<uint32_t>& string_prefix_length) {
+  const auto value_counts = AbstractHistogram<T>::_gather_value_distribution(segment);
+
+  if (value_counts.empty()) {
+    return nullptr;
+  }
+
+  const auto bins = EqualHeightHistogram<T>::_build_bins(value_counts, max_bin_count);
+
+  if constexpr (std::is_same_v<T, std::string>) {
+    const auto [characters, prefix_length] =  // NOLINT (Extra space before [)
+        get_default_or_check_string_histogram_prefix_settings(supported_characters, string_prefix_length);
+    return std::make_shared<EqualHeightHistogram<T>>(bins.bin_maximums, bins.bin_distinct_counts, bins.minimum,
+                                                     bins.total_count, characters, prefix_length);
+  } else {
+    DebugAssert(!supported_characters && !string_prefix_length,
+                "Do not provide string prefix prefix arguments for non-string histograms.");
+    return std::make_shared<EqualHeightHistogram<T>>(bins.bin_maximums, bins.bin_distinct_counts, bins.minimum,
+                                                     bins.total_count);
+  }
+}
+
+template <typename T>
+HistogramType EqualHeightHistogram<T>::histogram_type() const {
+  return HistogramType::EqualHeight;
+}
+
+template <typename T>
+std::string EqualHeightHistogram<T>::histogram_name() const {
+  return "EqualHeight";
+}
+
+template <typename T>
+BinID EqualHeightHistogram<T>::bin_count() const {
+  return _bin_maximums.size();
+}
+
+template <typename T>
+BinID EqualHeightHistogram<T>::_bin_for_value(const T value) const {
+  if (value < _minimum) {
+    return INVALID_BIN_ID;
+  }
+
+  const auto it = std::lower_bound(_bin_maximums.cbegin(), _bin_maximums.cend(), value);
+
+  if (it == _bin_maximums.cend()) {
+    return INVALID_BIN_ID;
+  }
+
+  return static_cast<BinID>(std::distance(_bin_maximums.cbegin(), it));
+}
+
+template <typename T>
+BinID EqualHeightHistogram<T>::_next_bin(const T value) const {
```

Comment:
Renamed it.

If the value is in between two bins (which can happen in an `EqualDistinctCountHistogram`), `_next_bin_for_value()` returns the index of the bin after the gap. If the value is in the last bin or larger than the maximum of the histogram, it returns `INVALID_BIN_ID`. Otherwise, it returns `_bin_for_value(value) + 1`.